### PR TITLE
Fix generated Java code for Thrift unions with multiple fields of same type

### PR DESCRIPTION
**Pull request by ryantimwilson**
Currently, when generating Java code for Thrift unions with multiple fields
of the same type, the code won't compile. This is because Swift generates
a constructor per field, so multiple constructors will have the same
arguments even though they initialize fields differently.

This patch changes the union template to use a builder paradigm in order
to allow for this case.

### CODE REVIEW

**Comment by siyengar** [comment]
@alandau could you take a look?

**Comment by alandau** [comment]
cc @SergeyMakarenko 

lgtm, except AFAIU this is a breaking change for code that used the constructor directly. Can you grep the relevant repos to make sure no such code exists and talk to the relevant people if it does?

File: swift-generator/src/main/resources/templates/java/common.st [hunk 1]
```diff
@@ -333,3 +336,47 @@ public String toString()
...
 }
 >>
+
+_union_constructor(context) ::= <<
+
+private Object value;
+private int id = -1;
+private String name;
+
+public <context.javaName>() {
```
Comment (ryandm) [inline swift-generator/src/main/resources/templates/java/common.st]: If possible, let's be more clear about what a valid state is for this object. If I understand correctly, this constructor allows a "nothing is set" instance to be created, where before this commit it's not possible to create a "nothing is set" instance.

In terms of consistency within this commit, the private constructor suggests that a valid state is `name != null`, but this constructor doesn't have that requirement. 

What's the real constraint we want to uphold?

Comment (ryantimwilson) [inline swift-generator/src/main/resources/templates/java/common.st]: When building, you only call setSomeField() once. For example, we don't want users to call builder.setSomeField().setAnotherField(). This condition is upheld below in the setters.

You're right, we should remove the condition check in the private constructor.

File: swift-generator/src/main/resources/templates/java/common.st [hunk 2]
```diff
@@ -333,3 +336,59 @@ public String toString()
...
+}
+
+<if(context.hasUniqueFieldTypes)>
```
Comment (alandau) [inline swift-generator/src/main/resources/templates/java/common.st]: This `if` (together with `getHasUniqueFieldTypes()`) is the only change since the prev version, right? In this case, lgtm

Comment (ryantimwilson) [inline swift-generator/src/main/resources/templates/java/common.st]: Pretty much, I did add a thrift file to test unions with and without duplicate types.

**Comment by ryantimwilson** [comment]
This was merged
